Update CI strategy matrix to use new RHEL 9 and RHEL 10 images

bthomee opened the pull request:
## High Level Overview of Change

This change uses the new RHEL 9 and 10 images to build and test the binary, and adds support for having different Docker image SHAs per distro-compiler combination.

### Context of Change

Instead of supporting RHEL each minor version, we are simplifying our pipelines by only supporting RHEL major versions. Our CI Docker images have already been updated accordingly, and we recently added support for RHEL 10 as well.

Up until now, the CI Docker images had all been rebuilt at the same time, but that is not necessarily true as the most recent push to the CI repo has shown where the RHEL images now have a different SHA than the Debian and Ubuntu ones.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

Bronek reviewed (approved):
Well done on using `6f723eb` and `0ab1e4c` where appropriate - I was for a minute "wait, why different commits" then I understood.

bthomee commented:
> Well done on using `6f723eb` and `0ab1e4c` where appropriate - I was for a minute "wait, why different commits" then I understood.

Yes - the `workflow_dispatch` on the separate distro workflow can also produce different SHAs per distro image based on whatever the last commit was to the default branch. Also, I toyed with the idea of having a SHA per distro, but I decided to implement it to allow each distro-compiler combination to have a different SHA; for instance, we might need to debug one specific image, and now we can push an image manually and reference it in an open PR.

bthomee committed:
refactor: Update CI strategy matrix to use new RHEL 9 and RHEL 10 images

This change uses the new RHEL 9 and 10 images to build and test the binary, and adds support for having different Docker image SHAs per distro-compiler combination.

Instead of supporting RHEL each minor version, we are simplifying our pipelines by only supporting RHEL major versions. Our CI Docker images have already been updated accordingly, and we recently added support for RHEL 10 as well. Up until now, the CI Docker images had all been rebuilt at the same time, but that is not necessarily true as the most recent push to the CI repo has shown where the RHEL images now have a different SHA than the Debian and Ubuntu ones.
